### PR TITLE
41 move changes from crypto3 zk up to commit 6a3101b4b89494df1f578e60dbb6de047fea8add

### PR DESCRIPTION
**Pull request by martun**
Changes are mostly related to fixes in LPC batching.

### CODE REVIEW

File: include/nil/actor/zk/commitments/batched_commitment.hpp [hunk 1]
```diff
@@ -91,7 +92,6 @@ namespace nil {
                     }
 
                     math::polynomial<typename field_type::value_type> get_U(std::size_t b_ind, std::size_t poly_ind) const {
-
                         const auto &points = _points.at(b_ind)[poly_ind];
                         BOOST_ASSERT(points.size() == this->_z.get_poly_points_number(b_ind, poly_ind));
                         std::vector<std::pair<typename field_type::value_type,typename field_type::value_type>> U_interpolation_points;
```
Comment (Zerg1996) [inline include/nil/actor/zk/commitments/batched_commitment.hpp:97]: I think we need #include<vector> in header and <utility> for pair.

File: include/nil/actor/zk/commitments/batched_commitment.hpp [hunk 2]
```diff
@@ -104,7 +104,26 @@ namespace nil {
...
-                    std::vector<std::vector<typename field_type::value_type>> get_unique_points_list() const{
+                    // We call them singles in recursive verifier
+                    std::vector<typename field_type::value_type> get_unique_points(){
+                        std::vector<typename field_type::value_type> result;
+//                        std::unordered_set<typename field_type::value_type> result_set;
+
+                        for( auto const &[k, point_batch]:_points ){
+                            for( auto const &point_set: point_batch ){
+                                for( auto const &point:point_set ){
+                                    if( std::find(result.begin(), result.end(), point) == result.end() ){
```
Comment (Zerg1996) [inline include/nil/actor/zk/commitments/batched_commitment.hpp:115]: Maybe let's use std::set for such type of insert? It's mark that we have unique_points.

File: include/nil/actor/zk/commitments/batched_commitment.hpp [hunk 3]
```diff
@@ -104,7 +104,26 @@ namespace nil {
...
+  //                                      result_set.insert(point);
+                                    }
+                                }
+                            }
+                        }
+
+                        return result;
+                    }
+
+                    std::vector<std::vector<typename field_type::value_type>> get_unique_point_sets_list() const{
```
Comment (Zerg1996) [inline include/nil/actor/zk/commitments/batched_commitment.hpp:126]: We can also use set here and don't need for loop inside to check was point in our list.

File: include/nil/actor/zk/commitments/detail/polynomial/basic_fri.hpp [hunk 4]
```diff
@@ -1023,4 +1018,4 @@ namespace nil {
     }            // namespace actor
 }    // namespace nil
 
-#endif    // ACTOR_ZK_COMMITMENTS_BASIC_FRI_HPP
+#endif    // ACTOR_ZK_COMMITMENTS_BASIC_FRI_HPP
```
Comment (Zerg1996) [inline include/nil/actor/zk/commitments/detail/polynomial/basic_fri.hpp:1021]: Missing newline at the end of file.

File: include/nil/actor/zk/commitments/polynomial/fri.hpp [hunk 5]
```diff
@@ -154,6 +154,8 @@ namespace nil {
                             evals_map, combined_U, combined_V,
                             transcript
                     );
+
+                    return true;
```
Comment (Zerg1996) [inline include/nil/actor/zk/commitments/polynomial/fri.hpp:158]: As I understand we didn't reach this code.

File: include/nil/actor/zk/commitments/polynomial/lpc.hpp [hunk 6]
```diff
@@ -242,33 +204,28 @@ namespace nil {
...
-                            for( auto const &it: this->_points){
-                                auto k = it.first;
-                                for (std::size_t i = 0; i < proof.z.get_batch_size(k); i++) {
-                                    combined_U[point_index] *= theta;
-                                    if (eval_map[k][i] == point_index) {
-                                        combined_U[point_index] += this->get_U(k, i);
-                                    }
+                        for (std::size_t p = 0; p < points.size(); p++){
+                            auto &point = points[p];
+                            bool first = true;
```
Comment (Zerg1996) [inline include/nil/actor/zk/commitments/polynomial/lpc.hpp:220]: Not sure we need this variable

File: include/nil/actor/zk/commitments/batched_commitment.hpp [hunk 7]
```diff
@@ -104,7 +104,26 @@ namespace nil {
                         return math::lagrange_interpolation(U_interpolation_points);
                     }
 
-                    std::vector<std::vector<typename field_type::value_type>> get_unique_points_list() const{
+                    // We call them singles in recursive verifier
+                    std::vector<typename field_type::value_type> get_unique_points(){
```
[the remark this answers is not in the document]
Comment (Zerg1996) [reply include/nil/actor/zk/commitments/batched_commitment.hpp:108]: Missing `const` in function declaration.

File: include/nil/actor/zk/snark/routing/as_waksman.hpp [hunk 8]
```diff
@@ -149,7 +149,7 @@ namespace nil {
                 /**
                  * Return the height of the AS-Waksman network's top sub-network.
                  */
-                std::size_t as_waksman_top_height(const std::size_t num_packets) {
+                inline std::size_t as_waksman_top_height(const std::size_t num_packets) {
```
Comment (Zerg1996) [inline include/nil/actor/zk/snark/routing/as_waksman.hpp:152]: We don't need const for std::size_t, missing const for function.

File: include/nil/actor/zk/snark/routing/as_waksman.hpp [hunk 9]
```diff
@@ -164,7 +164,7 @@ namespace nil {
                  *
                  * If top = true, return the top wire, otherwise return bottom wire.
                  */
-                std::size_t as_waksman_switch_output(size_t num_packets, std::size_t row_offset, std::size_t row_idx,
+                inline std::size_t as_waksman_switch_output(size_t num_packets, std::size_t row_offset, std::size_t row_idx,
                                                      bool use_top) {
                     std::size_t relpos = row_idx - row_offset;
                     assert(relpos % 2 == 0 && relpos + 1 < num_packets);
```
Comment (Zerg1996) [inline include/nil/actor/zk/snark/routing/as_waksman.hpp:170]: Maybe let's use assert with some message explanation?

File: include/nil/actor/zk/snark/routing/as_waksman.hpp [hunk 10]
```diff
@@ -164,7 +164,7 @@ namespace nil {
                  *
                  * If top = true, return the top wire, otherwise return bottom wire.
                  */
-                std::size_t as_waksman_switch_output(size_t num_packets, std::size_t row_offset, std::size_t row_idx,
+                inline std::size_t as_waksman_switch_output(size_t num_packets, std::size_t row_offset, std::size_t row_idx,
```
Comment (Zerg1996) [inline include/nil/actor/zk/snark/routing/as_waksman.hpp:167]: Const in function declaration missing

**Review by Zerg1996 — CHANGES_REQUESTED** [review]
Can you please check `const` declaration for every functions?

File: include/nil/actor/zk/commitments/batched_commitment.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (martun) [reply include/nil/actor/zk/commitments/batched_commitment.hpp:108]: As far as I understand this algorithm, we need to keep the order of points, when making them unique.

File: include/nil/actor/zk/commitments/batched_commitment.hpp (hunk 2 above)
Comment (martun) [reply include/nil/actor/zk/commitments/batched_commitment.hpp:115]: I suggested this change in crypto3-zk, it's not done in crypto3-zk yet, as you can see they created some commented out unordered_set<typename field_type::value_type> result_set;.

Will do the same thing here, once it's done in crypto3-zk.

File: include/nil/actor/zk/commitments/batched_commitment.hpp (hunk 1 above)
Comment (martun) [reply include/nil/actor/zk/commitments/batched_commitment.hpp:97]: Done.

File: include/nil/actor/zk/commitments/polynomial/fri.hpp (hunk 5 above)
Comment (martun) [reply include/nil/actor/zk/commitments/polynomial/fri.hpp:158]: Removed, thanks for this one!

File: include/nil/actor/zk/commitments/polynomial/lpc.hpp [hunk 11]
```diff
@@ -128,79 +128,41 @@ namespace nil {
...
-                                        combined_Q = Q_dfs;
-                                    } else {
-                                        combined_Q *= theta;
-                                        combined_Q += Q_dfs;
-                                    }
-                                }
-                            }
-                        } else {
+                        for (auto const &point: points){
                             bool first = true;
```
[the remark this answers is not in the document]
Comment (martun) [reply include/nil/actor/zk/commitments/polynomial/lpc.hpp:139]: Yes, removed.

File: include/nil/actor/zk/commitments/polynomial/lpc.hpp (hunk 6 above)
Comment (martun) [reply include/nil/actor/zk/commitments/polynomial/lpc.hpp:220]: Removed.

File: include/nil/actor/zk/snark/routing/as_waksman.hpp (hunk 9 above)
Comment (martun) [reply include/nil/actor/zk/snark/routing/as_waksman.hpp:170]: Will change this later.

File: include/nil/actor/zk/commitments/polynomial/lpc.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (martun) [reply include/nil/actor/zk/commitments/polynomial/lpc.hpp:139]: Removed.

**Review by Zerg1996 — APPROVED** [review]

**Pull request merged**